lib/ukcpio/cpio.c

```diff
@@ -58,6 +58,37 @@ int uk_syscall_r_utime(const char *, const struct utimbuf *);
 int uk_syscall_r_mkdir(const char *, mode_t);
 int uk_syscall_r_symlink(const char *, const char *);
 int uk_syscall_r_stat(const char *, struct stat *);
+int uk_syscall_r_unlinkat(int, const char *, int);
+int uk_syscall_r_rename(const char *, const char *);
+
+static int
+try_remove(const char *path)
+{
+	int r;
+
+	uk_pr_info("Trying to unlink %s\n", path);
+	r = uk_syscall_r_unlinkat(AT_FDCWD, path, 0);
+	if (!r || r == -ENOENT)
+		return 0;
+
+	uk_pr_info("Trying to rmdir %s\n", path);
+	r = uk_syscall_r_unlinkat(AT_FDCWD, path, AT_REMOVEDIR);
+	if (!r || r == -ENOENT) {
+		return 0;
+	} else {
+		char newpath[PATH_MAX];
+		char *newend = newpath + strlcpy(newpath, path, PATH_MAX);
+
+		if (newend - newpath + 2 > PATH_MAX) {
+			uk_pr_err("Cannot rename %s, path too long\n", path);
+			return -ENAMETOOLONG;
+		}
+		strcpy(newend, ".0");
```

Comment:
Can you also give a more detailed description (both as a comment and commit message) of what this helps us with and why it will not break future references to this directory? Something like:
```
The newly processed initrd has a directory that must replace a previously existing same name directory. If the previous directory is not empty, simply rename it so that it does not overlap with the new directory.
```
By the way, isn't memory allocated for the files within the renamed directory lost? Also what if some other thread is holding a reference to this directory or files within this directory? (`EBUSY`)

Are these changes only intended for the early initial mount?

Reply:
We do not rename directories because of any concern for their content, rather because `rm -rf` is annoyingly non-trivial to implement using POSIX APIs, and CPIO extraction is latency-critical for boot.

Ideally you would never get into a situation where the initrd replaces a directory with something else, but in the exceedingly rare case that does happen, we don't want to waste code & time recursively walking a dir tree unlinking things. Thus the rename.

> By the way, isn't memory allocated for the files within the renamed directory lost?

I don't see how that would happen, renamed directories don't lose their contents.

> Also what if some other thread is holding a reference to this directory or files within this directory? (EBUSY)

If you hold an open reference to a file, that file will not `EBUSY` when trying to unlink/rename it (at least not on *nix; Windows behaves like that), the two things are orthogonal. Indeed, an open reference keeps a file "alive" even after unlinking.

Reply:
> I don't see how that would happen, renamed directories don't lose their contents.

I understand. What I meant is they would remain allocated there but never be reachable.

> If you hold an open reference to a file

I meant the path of the file not an open file descriptor. The path would no longer be reachable.